master/buildbot/worker/gce.py

```diff
@@ -0,0 +1,249 @@
+        result = yield AbstractLatentWorker.reconfigService(self, name, password, **kwargs)
+        return result
+
+    @defer.inlineCallbacks
+    def getGCEService(self, sa_credentials):
+        gce = yield gceclientservice.GCEClientService.getService(
+            self.master, ['https://www.googleapis.com/auth/compute'], sa_credentials,
+            project=self.project, zone=self.zone, instance=self.instance,
+            renderer=self)
+        return gce
+
+    def resetClientService(self, gce):
+        self._gce = gce
+
+    def getMetadataFromState(self, state):
+        metadata = state['metadata']
+        dict = {}
+        for x in metadata.get('items', []):
+            dict[x['key']] = x['value']
+        return (metadata['fingerprint'], dict)
+
+    def getDesiredMetadata(self, build):
+        if ":" in self.masterFQDN:
+            host, port = self.masterFQDN.split(":")
+        else:
+            host = self.masterFQDN
+            port = 9989
+
+        return {
+            "WORKERNAME": self.name,
+            "WORKERPASS": self.password,
+            "BUILDMASTER": host,
+            "BUILDMASTER_PORT": port
+        }
+
+    def updateMetadata(self, build, metadata):
+        result = metadata.copy()
+        for key in GCE_NODE_METADATA_KEYS:
+            result.pop(key, None)
+        result.update(self.getDesiredMetadata(build))
+        return result
+
+    def getMetadataItemsFromDict(self, metadata):
+        return [{"key": key, "value": metadata[key]} for key in metadata]
+
+    def setMetadata(self, fingerprint, metadata):
+        return self._gce.setMetadata(fingerprint=fingerprint,
+            items=self.getMetadataItemsFromDict(metadata))
+
+    def getCurrentDiskName(self, disks):
+        for d in disks:
+            if d['boot']:
+                return d['deviceName']
+
+    def getNewDiskName(self, metadata):
+        gen = int(metadata.get(DISK_NAME_GEN_KEY, 0)) + 1
+        metadata[DISK_NAME_GEN_KEY] = str(gen)
+        return "{0}-{1}".format(self.instance, gen)
+
+    def createBootDisk(self, disk_name):
+        if self.useSSD:
+            diskType = "pd-ssd"
+        else:
+            diskType = "pd-standard"
+
+        return self._gce.createDisk(image=self.image, name=disk_name, type=diskType)
+
+    @defer.inlineCallbacks
+    def start_instance(self, build):
+        instance_state = yield self._gce.getInstanceState()
+
+        fingerprint, metadata = self.getMetadataFromState(instance_state)
+        updated_metadata = self.updateMetadata(build, metadata)
+
+        instance_stop = None
+        if instance_state['status'] not in ('STOPPED', 'TERMINATED'):
+            if self.stopInstanceOnStop:
+                # Should not have happened ... warn the user
+                log.warn(("gce: {0} is unexpectedly running. Either buildbot failed to stop it "
+                    "on the last stop, or the node was started externally. Requesting stop").format(self.instance))
```

Comment:
Looks like some lines are too long, they overflow even the github web interface. Could you wrap them, preferably to 80 or 100 chars?